test/shared/src/main/scala/zio/test/ZTestLogger.scala

```diff
@@ -0,0 +1,118 @@
+package zio.test
+
+import zio._
+
+import scala.annotation.tailrec
+
+/**
+ * A `ZTestLogger` is an implementation of a `ZLogger` that writes all log
+ * messages to an internal data structure. The contents of this data structure
+ * can be accessed using the `logOutput` operator. This makes it easy to write
+ * tests to verify that expected messages are being logged.
+ *
+ * {{{
+ * test("logging works") {
+ *   for {
+ *     _      <- ZIO.logDebug("It's alive!")
+ *     output <- ZTestLogger.logOutput
+ *   } yield assertTrue(output.length == 1) &&
+ *     assertTrue(output(0).message() == "It's alive!") &&
+ *     assertTrue(output(0).logLevel == LogLevel.Debug)
+ * }
+ * }}}
+ */
+sealed trait ZTestLogger[-Message, +Output] extends ZLogger[Message, Output] {
+
+  /**
+   * Returns the contents of the log.
+   */
+  def logOutput: UIO[Chunk[ZTestLogger.LogEntry]]
+}
+
+object ZTestLogger {
+
+  /**
+   * A layer which constructs a new `ZTestLogger` and runs the effect it is
+   * provided to with the `RuntimeConfig` updated to add the `ZTestLogger`.
+   */
+  val default: ZLayer[Any, Nothing, Any] =
+    ZLayer {
+      for {
+        runtimeConfig              <- ZManaged.runtimeConfig
+        testLoggers                <- ZTestLogger.make.toManaged
+        (stringLogger, causeLogger) = testLoggers
+        runtimeConfigAspect         = RuntimeConfigAspect.addLogger(stringLogger) >>> RuntimeConfigAspect.addLogger(causeLogger)
+        _                          <- ZManaged.withRuntimeConfig(runtimeConfigAspect(runtimeConfig))
+      } yield ()
+    }
+
+  /**
+   * Accesses the contents of the current test logger.
+   */
+  val logOutput: UIO[Chunk[ZTestLogger.LogEntry]] =
+    ZIO.runtimeConfig.flatMap { runtimeConfig =>
+      runtimeConfig.loggers
+        .getAll[String]
+        .collectFirst { case testLogger: ZTestLogger[_, _] =>
+          testLogger.logOutput
+        }
+        .getOrElse(ZIO.dieMessage("Defect: ZTestLogger is missing"))
+    }
+
+  /**
+   * A log entry captures all of the contents of a log message as a data
+   * structure.
+   */
+  final case class LogEntry(
+    trace: ZTraceElement,
+    fiberId: FiberId,
+    logLevel: LogLevel,
+    message: () => String,
```

Comment:
I think it may be nicer if the logged `Cause`s are not pretty printed but stored as is, so message could be an `Exit[Any, String`.

Reply:
It seems like that might make the common case less ergonomic, because then you don't have the guarantee that you always have a `String` message so it is harder to make assertions about it. Potentially we could keep track of the string logs and cause logs separately and have a separate operator to access them. May also depend on what we decide to do about the overall relationship between `String` and `Cause` logging.